File: src/leetcode/0094_binary_tree_inorder_traversal.rs
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None
    }
  }
}

use std::rc::Rc;
use std::cell::RefCell;
//impl Solution {
    pub fn inorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        match root {
            Some(node) => {
                let mut result: Vec<i32> = Vec::new();
                // left traversal
                let mut lft: Vec<i32> = if let Some(l) = &node.borrow().left {
                    Self::inorder_traversal(Some(Rc::clone(&l)))
                } else {
                    Vec::new()
                };
                result.append(&mut lft);

                result.push(node.borrow().val);
                
                // right traversal
                let mut rgt = if let Some(r) = &node.borrow().right {
                    Self::inorder_traversal(Some(Rc::clone(&r)))
                } else {
                    Vec::new()
                };
                result.append(&mut rgt);
                result
            },
            None => { return Vec::new(); }
        }
    }
//}

File: src/leetcode/0094_binary_tree_inorder_traversal.py
# Definition for a binary tree node.
from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        ans = []

        def recursive(node):
            if not node:
                return
            recursive(node.left)
            ans.append(node.val)
            recursive(node.right)

        recursive(root)
        return ans

    def inorder_traversal_stack(self, root):
        ans = []
        visited = set()
        stack = [root]
        while stack:
            node = stack.pop()
            if not node:
                continue
            if node in visited:
                ans.append(node.val)
                continue
            visited.add(node)
            stack.append(node.right)
            stack.append(node)
            stack.append(node.left)
        return ans
